src/core/thread/network_data_publisher.hpp

```diff
@@ -0,0 +1,464 @@
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definition of Network Data Publisher.
+ */
+
+#ifndef NETWORK_DATA_PUBLISHER_HPP_
+#define NETWORK_DATA_PUBLISHER_HPP_
+
+#include "openthread-core-config.h"
+
+#if OPENTHREAD_CONFIG_NETDATA_PUBLISHER_ENABLE
+
+#if !OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE && !OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE
+#error "OPENTHREAD_CONFIG_NETDATA_PUBLISHER_ENABLE requires either OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE"\
+            "or OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE"
+#endif
+
+#include <openthread/netdata_publisher.h>
+
+#include "common/clearable.hpp"
+#include "common/equatable.hpp"
+#include "common/error.hpp"
+#include "common/locator.hpp"
+#include "common/non_copyable.hpp"
+#include "common/notifier.hpp"
+#include "common/string.hpp"
+#include "common/timer.hpp"
+#include "net/ip6_address.hpp"
+#include "thread/network_data_types.hpp"
+
+namespace ot {
+namespace NetworkData {
+
+/**
+ * This class implements the Network Data Publisher.
+ *
+ * It provides mechanisms to limit the number of similar Service and/or Prefix (on-mesh prefix or external route)
+ * entries in the Thread Network Data by monitoring the Network Data and managing if or when to add or remove entries.
+ *
+ */
+class Publisher : public InstanceLocator, private NonCopyable
+{
+    friend class ot::Notifier;
+
+public:
+    /**
+     * This enumeration represents the events reported from the Publisher callbacks.
+     *
+     */
+    enum Event : uint8_t
+    {
+        kEventEntryAdded   = OT_NETDATA_PUBLISHER_EVENT_ENTRY_ADDED,   ///< Entry is added to Network Data.
+        kEventEntryRemoved = OT_NETDATA_PUBLISHER_EVENT_ENTRY_REMOVED, ///< Entry is removed from Network Data.
+    };
+
+    /**
+     * This constructor initializes `Publisher` object.
+     *
+     * @param[in]  aInstance     A reference to the OpenThread instance.
+     *
+     */
+    explicit Publisher(Instance &aInstance);
+
+#if OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE
```

Comment:
We cannot not use `OPENTHREAD_CONFIG_SRP_SERVER_ENABLE` here, since it is possible for user to use `Publisher` without actually using the OT SRP server implementation. For example, in the case where SRP server is provided by infrastructure.

Reply:
Ok, I see. Make sense